File: ios/exelbid_plugin/Sources/exelbid_plugin/Mediation/MediatedInterstitialAdHandle.swift
import ExelBidSDK
import Flutter
import UIKit

/// Per-instance handle for `ExelBidSDK.EBMediatedInterstitialAd`. Mirrors
/// `InterstitialAdHandle`, adding waterfall events and the winning network
/// (reported with `onLoad`).
final class MediatedInterstitialAdHandle: NSObject, FlutterStreamHandler {

    private let id: String
    private let interstitialAd: EBMediatedInterstitialAd
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?

    init(id: String, adUnitId: String, options: [String: Any]?,
         perNetworkTimeout: Double?, messenger: FlutterBinaryMessenger) {
        self.id = id
        self.interstitialAd = EBMediatedInterstitialAd(adUnitId: adUnitId)
        self.interstitialAd.options = AdOptionsMapper.decode(options)
        if let perNetworkTimeout = perNetworkTimeout {
            self.interstitialAd.perNetworkTimeout = perNetworkTimeout
        }
        self.methodChannel = FlutterMethodChannel(
            name: ChannelNames.MediatedInterstitial.method(forId: id),
            binaryMessenger: messenger
        )
        self.eventChannel = FlutterEventChannel(
            name: ChannelNames.MediatedInterstitial.events(forId: id),
            binaryMessenger: messenger
        )

        super.init()

        interstitialAd.rootViewControllerProvider = { TopViewController.find() }
        wireCallbacks()
        eventChannel.setStreamHandler(self)
        methodChannel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
    }

    deinit {
        methodChannel.setMethodCallHandler(nil)
        eventChannel.setStreamHandler(nil)
    }

    // MARK: - FlutterStreamHandler

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // MARK: - Methods

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "load":
            interstitialAd.load()
            result(nil)

        case "present":
            guard let controller = TopViewController.find() else {
                result(FlutterError(
                    code: "no_presenter",
                    message: "Could not find a UIViewController to present from.",
                    details: nil
                ))
                return
            }
            Task { @MainActor [interstitialAd] in
                interstitialAd.present(from: controller)
            }
            result(nil)

        case "isReady":
            result(interstitialAd.isReady)

        case "stop":
            interstitialAd.stop()
            result(nil)

        case "dispose":
            interstitialAd.stop()
            InstanceRegistry.shared.remove(id: id)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Callback wiring

    private func wireCallbacks() {
        interstitialAd.onLoad = { [weak self, weak interstitialAd] in
            var payload: [String: Any] = ["event": "onLoad"]
            if let network = interstitialAd?.winningNetwork {
                payload["winningNetwork"] = network
            }
            self?.send(payload)
        }
        interstitialAd.onFailureBlock = { [weak self] error in
            self?.send([
                "event": "onFail",
                "error": AdErrorMapper.encode(error as NSError),
            ])
        }
        interstitialAd.onWaterfallEvent = { [weak self] event in
            self?.send([
                "event": "onWaterfall",
                "waterfall": WaterfallEventMapper.encode(event),
            ])
        }
        interstitialAd.onWillAppear    = { [weak self] in self?.send(["event": "onWillAppear"]) }
        interstitialAd.onDidAppear     = { [weak self] in self?.send(["event": "onDidAppear"]) }
        interstitialAd.onWillDisappear = { [weak self] in self?.send(["event": "onWillDisappear"]) }
        interstitialAd.onDidDisappear  = { [weak self] in self?.send(["event": "onDidDisappear"]) }
        interstitialAd.onClick         = { [weak self] in self?.send(["event": "onClick"]) }
        interstitialAd.onLeaveApp      = { [weak self] in self?.send(["event": "onLeaveApp"]) }
        interstitialAd.onClickFinish   = { [weak self] in self?.send(["event": "onClickFinish"]) }
    }

    private func send(_ payload: [String: Any]) {
        if Thread.isMainThread {
            eventSink?(payload)
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.eventSink?(payload)
            }
        }
    }
}

File: android/src/main/kotlin/com/exelbid/flutter/mediation/MediatedInterstitialAdHandle.kt
package com.exelbid.flutter.mediation

import android.app.Activity
import android.content.Context
import com.exelbid.flutter.ChannelNames
import com.exelbid.flutter.InstanceRegistry
import com.exelbid.flutter.mappers.AdErrorMapper
import com.exelbid.flutter.mediation.adapter.InterstitialMediationAdapter
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Per-instance handle for a mediated interstitial. Owns a per-id MethodChannel
 * (load/isReady/present/stop/dispose) + EventChannel, and a
 * [WaterfallOrchestrator] that runs the INTERSTITIAL waterfall.
 *
 * Two-stage: `load()` runs the waterfall (winner becomes ready), `present()`
 * shows the winner on the current Activity. Lifecycle/interaction events flow
 * through the orchestrator's adapter callback to this handle's sink.
 *
 * Held by [InstanceRegistry] until `dispose`. Mirrors iOS `InterstitialAdHandle`.
 */
class MediatedInterstitialAdHandle(
    private val id: String,
    private val adUnitId: String,
    private val options: Map<String, Any?>?,
    private val perNetworkTimeoutMs: Long?,
    private val applicationContext: Context,
    messenger: BinaryMessenger,
    private val activityProvider: () -> Activity?,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val methodChannel =
        MethodChannel(messenger, ChannelNames.MediatedInterstitial.method(id))
    private val eventChannel =
        EventChannel(messenger, ChannelNames.MediatedInterstitial.events(id))

    private var sink: EventChannel.EventSink? = null
    private val pending = ArrayList<Map<String, Any?>>()

    private var orchestrator: WaterfallOrchestrator? = null
    private var adapter: InterstitialMediationAdapter? = null
    private var ready = false

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "load" -> {
                load()
                result.success(null)
            }
            "isReady" -> result.success(ready && adapter?.isReady() == true)
            "present" -> {
                present()
                result.success(null)
            }
            "stop" -> {
                stop()
                result.success(null)
            }
            "dispose" -> {
                dispose()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun load() {
        orchestrator?.dispose()
        adapter = null
        ready = false

        // Interstitials prefer an Activity context (ExelBid.show() uses it).
        val loadContext: Context = activityProvider() ?: applicationContext

        val orch = WaterfallOrchestrator(
            format = AdFormat.INTERSTITIAL,
            unitId = adUnitId,
            context = loadContext,
            activity = activityProvider(),
            options = options,
            perNetworkTimeoutMs = perNetworkTimeoutMs,
            emit = ::emit,
            onWon = { won, _ ->
                adapter = won as? InterstitialMediationAdapter
                ready = adapter != null
            },
        )
        orchestrator = orch
        orch.start()
    }

    private fun present() {
        val a = adapter
        if (a == null || !a.isReady()) {
            emit(mapOf("event" to "onFail", "error" to AdErrorMapper.encode(9, "Interstitial not ready")))
            return
        }
        val activity = activityProvider()
        if (activity == null) {
            emit(mapOf("event" to "onFail", "error" to AdErrorMapper.encode(9, "No activity to present interstitial")))
            return
        }
        emit(mapOf("event" to "onWillAppear"))
        a.show(activity)
    }

    private fun stop() {
        orchestrator?.dispose()
        orchestrator = null
        adapter = null
        ready = false
    }

    private fun dispose() {
        stop()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        sink = null
        pending.clear()
        InstanceRegistry.remove(id)
    }

    /** Invoked on the main thread (orchestrator marshals; method calls are main). */
    private fun emit(event: Map<String, Any?>) {
        val s = sink
        if (s != null) s.success(event) else pending.add(event)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        if (events != null && pending.isNotEmpty()) {
            pending.forEach { events.success(it) }
            pending.clear()
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}
